File: js/src/mcpserve.ts
/**
 * MCP — INBOUND: serve a toolkit as an MCP server (SPEC §7C).
 *
 * The inbound mirror of the A2A `serve` (§7B). Where A2A advertises the toolkit's
 * **skills** and fulfils a Task through the whole client loop, MCP advertises the
 * toolkit's **unified tools** (every source — mcp · skill · native · http ·
 * builtin · a2a) and dispatches each `tools/call` straight to `Tool.execute`. The
 * calling MCP client *is* the LLM host, so there is **no client, no Task, and no
 * TaskStore** — just `tools/list` + `tools/call` over the toolkit registry.
 *
 * Transport: **streamable-HTTP**, mounted at `POST /mcp` on the `serve(addr, …)`
 * server alongside the A2A routes (see serve.ts), built on the same
 * `@modelcontextprotocol/sdk` the client side already uses.
 */
import { Server } from "@modelcontextprotocol/sdk/server/index.js"
import { ListToolsRequestSchema, CallToolRequestSchema, McpError, ErrorCode } from "@modelcontextprotocol/sdk/types.js"
import type { Tool, ToolSource, ToolResult } from "./types.js"
import type { ContentPart } from "./content.js"

const DEFAULT_NAME = "toolnexus"
const DEFAULT_VERSION = "0.1.0"

/** Opt-in MCP serve profile — the toolkit is exposed as an MCP server. */
export interface MCPServeConfig {
  /** Advertised server name (initialize `serverInfo.name`). Default "toolnexus". */
  name?: string
  /** Advertised server version. Default "0.1.0". */
  version?: string
  /** Subset of tool names to expose; omit ⇒ all. Unknown names are ignored. */
  tools?: string[]
}

/** Surfaced on each inbound `tools/call`. */
export interface OnCallEvent {
  name: string
  source: ToolSource
  ms: number
  isError: boolean
}

export type OnCall = (ev: OnCallEvent) => void | Promise<void>

/**
 * The tools this profile exposes: every toolkit tool, filtered to `cfg.tools`
 * when set (unknown names in the filter are ignored, never an error).
 */
export function exposedMcpTools(tools: Tool[], cfg?: MCPServeConfig): Tool[] {
  if (!cfg?.tools) return tools
  const wanted = new Set(cfg.tools)
  return tools.filter((t) => wanted.has(t.name))
}

/**
 * Build a low-level MCP `Server` exposing `tools` as MCP tools. `tools/list`
 * maps each Tool's name (verbatim — already sanitized at registration),
 * description, and inputSchema; `tools/call` dispatches to `Tool.execute` and
 * maps the `ToolResult` to a `CallToolResult` (`output` → text, `isError`
 * propagates). A fresh server is cheap to build, so the HTTP path makes one per
 * request (stateless).
 */
export function buildMcpServer(tools: Tool[], cfg?: MCPServeConfig, onCall?: OnCall): Server {
  const byName = new Map(tools.map((t) => [t.name, t]))
  const server = new Server(
    { name: cfg?.name ?? DEFAULT_NAME, version: cfg?.version ?? DEFAULT_VERSION },
    { capabilities: { tools: {} } },
  )

  server.setRequestHandler(ListToolsRequestSchema, async () => ({
    tools: tools.map((t) => ({ name: t.name, description: t.description, inputSchema: t.inputSchema })),
  }))

  server.setRequestHandler(CallToolRequestSchema, async (req, extra) => {
    const tool = byName.get(req.params.name)
    if (!tool) throw new McpError(ErrorCode.InvalidParams, `Unknown tool: ${req.params.name}`)
    const started = Date.now()
    // The call is a single synchronous tool invocation honoring the MCP
    // request's cancellation. An execute() throw becomes an isError result —
    // never a crash.
    let result: ToolResult
    try {
      result = await tool.execute(req.params.arguments ?? {}, { signal: extra.signal })
    } catch (e) {
      result = { output: e instanceof Error ? e.message : String(e), isError: true }
    }
    if (onCall) {
      try {
        await onCall({ name: tool.name, source: tool.source, ms: Date.now() - started, isError: result.isError })
      } catch {
        // Host callback errors are isolated.
      }
    }
    return { content: [{ type: "text", text: result.output }, ...partBlocks(result.parts)], isError: result.isError }
  })

  return server
}

/**
 * §7C. A served tool's `ToolResult.parts` become MCP content blocks appended after the
 * text part, in order — image ⇒ image content, audio ⇒ audio content, file ⇒ an embedded
 * resource (a url-backed file keeps its uri, a byte-backed one carries the blob).
 */
function partBlocks(parts: ContentPart[] | undefined): any[] {
  if (!parts?.length) return []
  const out: any[] = []
  for (const p of parts) {
    if (p.type === "text") out.push({ type: "text", text: p.text })
    else if (p.type === "image" && p.data) out.push({ type: "image", data: p.data, mimeType: p.mimeType })
    else if (p.type === "audio" && p.data) out.push({ type: "audio", data: p.data, mimeType: p.mimeType })
    else {
      const uri = p.url ?? `file:///${(p as any).name ?? "resource"}`
      const resource: any = { uri, mimeType: p.mimeType }
      if (p.data) resource.blob = p.data
      out.push({ type: "resource", resource })
    }
  }
  return out
}

File: golang/mcpserve.go
// MCP — INBOUND: serve a toolkit as an MCP server (SPEC §7C).
//
// The inbound mirror of the A2A Serve (§7B). Where A2A advertises the toolkit's
// SKILLS and fulfils a Task through the whole client loop, MCP advertises the
// toolkit's UNIFIED tools (every source — mcp · skill · native · http · builtin ·
// a2a) and dispatches each tools/call straight to Tool.Execute. The calling MCP
// client IS the LLM host, so there is no client, no Task, and no TaskStore — just
// tools/list + tools/call over the toolkit registry.
//
// Transport — streamable-HTTP, built on the same github.com/mark3labs/mcp-go the
// client side already uses: mounted at POST /mcp on the Serve(addr, …) server,
// alongside the A2A routes (see serve.go).

package toolnexus

import (
	"context"
	"encoding/json"
	"time"

	"github.com/mark3labs/mcp-go/mcp"
	"github.com/mark3labs/mcp-go/server"
)

const (
	mcpServeDefaultName    = "toolnexus"
	mcpServeDefaultVersion = "0.1.0"
)

// MCPServeConfig is the opt-in MCP serve profile — the toolkit is exposed as an
// MCP server.
type MCPServeConfig struct {
	// Name is the advertised server name (initialize serverInfo.name). Default
	// "toolnexus".
	Name string `json:"name,omitempty"`
	// Version is the advertised server version. Default "0.1.0".
	Version string `json:"version,omitempty"`
	// Tools is the subset of tool names to expose; nil ⇒ all. Unknown names in
	// the filter are ignored (never an error).
	Tools []string `json:"tools,omitempty"`
}

// OnCallEvent is surfaced on each inbound tools/call.
type OnCallEvent struct {
	Name    string
	Source  ToolSource
	Ms      int64
	IsError bool
}

// OnCall is the per-call callback passed to Serve.
type OnCall func(ev OnCallEvent)

// ExposedMcpTools returns the tools this profile exposes: every toolkit tool,
// filtered to cfg.Tools when set (unknown names in the filter are ignored, never
// an error).
func ExposedMcpTools(tools []Tool, cfg *MCPServeConfig) []Tool {
	if cfg == nil || cfg.Tools == nil {
		return tools
	}
	wanted := make(map[string]bool, len(cfg.Tools))
	for _, n := range cfg.Tools {
		wanted[n] = true
	}
	out := make([]Tool, 0, len(tools))
	for _, t := range tools {
		if wanted[t.Name] {
			out = append(out, t)
		}
	}
	return out
}

// buildMcpServer builds a mark3labs MCPServer exposing tools as MCP tools.
// tools/list maps each Tool's name (verbatim — already sanitized at
// registration), description, and inputSchema; tools/call dispatches to
// Tool.Execute and maps the ToolResult to a CallToolResult (output → text,
// isError propagates). A fresh server is cheap to build, so the HTTP path makes
// one per server instance (stateless).
func buildMcpServer(tools []Tool, cfg *MCPServeConfig, onCall OnCall) *server.MCPServer {
	name := mcpServeDefaultName
	version := mcpServeDefaultVersion
	if cfg != nil {
		if cfg.Name != "" {
			name = cfg.Name
		}
		if cfg.Version != "" {
			version = cfg.Version
		}
	}

	srv := server.NewMCPServer(name, version, server.WithToolCapabilities(false))
	for _, t := range tools {
		srv.AddTool(mcpToolFor(t), mcpHandlerFor(t, onCall))
	}
	return srv
}

// mcpToolFor builds the wire tool definition for t. The name is used verbatim
// (already sanitized at registration — NOT re-sanitized) and the inputSchema is
// the tool's parameters JSON Schema, passed through as a raw schema so it is
// advertised byte-for-byte.
func mcpToolFor(t Tool) mcp.Tool {
	schema := t.InputSchema
	if schema == nil {
		schema = JSONSchema{}
	}
	raw, err := json.Marshal(schema)
	if err != nil {
		raw = []byte("{}")
	}
	return mcp.NewToolWithRawSchema(t.Name, t.Description, raw)
}

// mcpHandlerFor wraps Tool.Execute as an MCP tool handler. An Execute error (or a
// non-nil error return) becomes an isError result — never a protocol crash. The
// onCall callback fires per call and its own errors/panics are isolated.
func mcpHandlerFor(t Tool, onCall OnCall) server.ToolHandlerFunc {
	return func(ctx context.Context, req mcp.CallToolRequest) (*mcp.CallToolResult, error) {
		started := time.Now()

		var tctx *ToolContext
		if ctx != nil {
			tctx = &ToolContext{Ctx: ctx}
		}
		res, err := t.Execute(req.GetArguments(), tctx)
		if err != nil {
			res = ToolResult{Output: err.Error(), IsError: true}
		}

		if onCall != nil {
			func() {
				defer func() { _ = recover() }()
				onCall(OnCallEvent{
					Name:    t.Name,
					Source:  t.Source,
					Ms:      time.Since(started).Milliseconds(),
					IsError: res.IsError,
				})
			}()
		}

		out := mcp.NewToolResultText(res.Output)
		out.IsError = res.IsError
		return out, nil
	}
}
